tls/s2n_server_hello_retry.c

```diff
@@ -32,6 +32,15 @@ uint8_t hello_retry_req_random[S2N_TLS_RANDOM_DATA_LEN] = {
     0xC2, 0xA2, 0x11, 0x16, 0x7A, 0xBB, 0x8C, 0x5E, 0x07, 0x9E, 0x09, 0xE2, 0xC8, 0xA8, 0x33, 0x9C
 };
 
+bool s2n_is_hello_retry_valid(struct s2n_connection *conn)
+{
+    bool has_correct_server_version = conn->server_protocol_version >= S2N_TLS13;
+    bool has_correct_client_version = conn->client_protocol_version == S2N_TLS13;
+    bool has_correct_random = (memcmp(hello_retry_req_random, conn->secure.server_random, S2N_TLS_RANDOM_DATA_LEN) == 0);
+
+    return has_correct_server_version && has_correct_client_version && has_correct_random;
+}
```

Comment:
Are you shooting for constant time here? (I don't think so). Usually we'd write this as ...

    if (conn->server_protocol_version < S2N_TLS13) {
        return false;
    }

    if (conn->client_protocol_version != S2N_TLS13) {
        return false;
    }

    if (memcmp(hello_retry_req_random, onn->secure.server_random, S2N_TLS_RANDOM_DATA_LEN)) {
        return false;
    }

    return true;

Separately; why is it exact match for the client version, but higher-than match for the server version?

Reply:
Sounds good, I'll make the change suggested. Re:  '>=" match, this is because we are yet to completely resolve this issue: https://github.com/awslabs/s2n/issues/1681

Reply:
Another way to clean this up is to _validate_ the state instead of returning if it's valid or not:

```c
int s2n_hello_retry_validate(struct s2n_connection *conn)
{
    nonnull_check(conn);
    ENSURE_POSIX(conn->server_protocol_version >= S2N_TLS13, S2N_ERR_INVALID_ARGUMENT);
    ENSURE_POSIX(conn->client_protocol_version == S2N_TLS13, S2N_ERR_INVALID_ARGUMENT);
    ENSURE_POSIX(memcmp(hello_retry_req_random, conn->secure.server_random, S2N_TLS_RANDOM_DATA_LEN) == 0);

    return S2N_SUCCESS;
}
```

and then callers would call with

```c
s2n_hello_retry_validate(conn) == S2N_SUCCESS
```